Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp

```diff
@@ -546,6 +550,35 @@ static ExceptionOr<void> validate(const ApplePayModifier& applePayModifier)
     }
 #endif
 
+#if ENABLE(APPLE_PAY_DEFERRED_PAYMENTS)
+    if (auto& deferredPaymentRequest = applePayModifier.deferredPaymentRequest) {
+        auto& deferredBilling = deferredPaymentRequest->deferredBilling;
+        if (deferredBilling.paymentTiming != ApplePayPaymentTiming::Deferred)
+            return Exception(TypeError, "'deferredBilling' must be a 'deferred' line item."_s);
+        if (!deferredBilling.label)
+            return Exception(TypeError, "Missing label for 'deferredBilling'."_s);
+        if (!isValidDecimalMonetaryValue(deferredBilling.amount) && deferredBilling.type != ApplePayLineItem::Type::Pending)
+            return Exception(TypeError, makeString('"', deferredBilling.amount, "\" is not a valid amount."));
+
+        if (auto& managementURL = deferredPaymentRequest->managementURL; !URL { managementURL }.isValid())
+            return Exception(TypeError, makeString('"', managementURL, "\" is not a valid URL."));
+
+        if (auto& freeCancellationDate = deferredPaymentRequest->freeCancellationDate; std::isnan(freeCancellationDate)) {
+            if (auto& freeCancellationDateTimeZone = deferredPaymentRequest->freeCancellationDateTimeZone; !freeCancellationDateTimeZone.isEmpty())
+                return Exception(TypeError, "Unexpected 'freeCancellationDateTimeZone' when 'freeCancellationDate' is not set."_s);
+        } else if (auto& freeCancellationDateTimeZone = deferredPaymentRequest->freeCancellationDateTimeZone; freeCancellationDateTimeZone.isEmpty())
+            return Exception(TypeError, "Missing 'freeCancellationDateTimeZone' when 'freeCancellationDate' is set."_s);
+        else if (!CFTimeZoneCreateWithName(nullptr, freeCancellationDateTimeZone.createCFString().get(), false))
+            return Exception(TypeError, makeString('"', freeCancellationDateTimeZone, "\" is not a valid time zone."));
+
+        if (auto& paymentDescription = deferredPaymentRequest->paymentDescription; paymentDescription.isEmpty())
+            return Exception(TypeError, "Missing 'paymentDescription'."_s);
+
+        if (auto& tokenNotificationURL = deferredPaymentRequest->tokenNotificationURL; !tokenNotificationURL.isNull() && !URL { tokenNotificationURL }.isValid())
+            return Exception(TypeError, makeString('"', tokenNotificationURL, "\" is not a valid URL."));
+    }
```

Comment:
There seems to be a lot of duplication between this and `convertAndValidate`. Can we share the code?

Reply:
Agreed.

Reply:
Ok, I've managed to combine them.
I tried to put them in a new file "ApplePayDeferredPaymentRequest.cpp", but the unified build got confused -- please don't make me fix these things today! Happy to file a follow-up radar, which should also try to de-duplicate the validation of other structs.

Reply:
Okay, we can do that in a follow-up.

Reply:
I've filed rdar://106207791 (De-duplicate payment validation code)